### DISCUSSION TITLE
Issue with modulators.

### DISCUSSION THREAD

**Discussion by bradholland** [FluidSynth/fluidsynth#1453] [Q&A] [answered]
Hi everyone,

Apologies if this has been asked before!

I’m working on an LV2 plugin that wraps FluidSynth in a headless setup. I’ve added LV2 control ports for parameters like Filter Cutoff, Resonance, Volume Attack, Decay, Sustainand Release to introduce variation to the sounds.

I’ve encountered some challenges with getting the modulators to respond consistently:
	1.	When I use MIDI CCs to target these modulators, they work—but only if the modulators are pre-defined in the SoundFont itself.
	2.	I tried directly setting fluid_gen parameters (e.g., GEN_FILTERFC, GEN_FILTERQ, etc.), which seemed to work but only affected a single voice of a preset while leaving others unchanged.

Am I correct in thinking that the only way to achieve a “global” filter or amp envelope across all voices is to pre-bake the MIDI CC modulators into the SoundFont? Or is there a more efficient way to handle this?

Thanks so much for your time and help!

**Comment by derselbst** [comment, marked as answer]
> Am I correct in thinking that the only way to achieve a “global” filter or amp envelope across all voices is to pre-bake the MIDI CC modulators into the SoundFont?

This is one possible option and it is the standard compliant way to achieve this in the Soundfont world. It is the only portable approach across SF2 compliant synth's.

> Or is there a more efficient way to handle this?

Since you're trying to set up custom default modulators, there is a fluidsynth specific API call for this purpose. Pls. refer to https://github.com/FluidSynth/fluidsynth/wiki/FluidFeatures

Setting fluid_gen parameters directly is not appropriate for your use-case.

**Comment by bradholland** [comment]
appreciate the guidance, thank you.
